### ISSUE TITLE
ViT from scratch

### ISSUE THREAD

**Issue by danjac94** [IBM/terratorch#634]
Need some help for running the ViT from scratch with Upernet which will be the correct configuration?

model:
  class_path: terratorch.tasks.SemanticSegmentationTask
  init_args:
    model_args:
      decoder: UperNetDecoder
      backbone_pretrained: False
      backbone: vit_small_patch16_224
      backbone_img_size: 256
      decoder_scale_modules: True
      # decoder_in_channels: 1024
      decoder_channels: 256
      # backbone_in_channels: 6
      backbone_model_bands:
        - BLUE
        - GREEN
        - RED
        - NIR_NARROW
        - SWIR_1
        - SWIR_2
      # backbone_out_indices:
          # - 5
          # - 11
          # - 17
          # - 23
      # num_frames: 1
      num_classes: 5
      head_dropout: 0.1
      head_channel_list:
        - 256
      necks:
        - name: SelectIndices
#          indices: [2, 5, 8, 11] # 100M models
#          indices: [5, 11, 17, 23] # 300M models
#          indices: [7, 15, 23, 31] # 600M models
        - name: ReshapeTokensToImage
    loss: ce
    class_weights:
      - 1.2232142857142858
      - 1.2232142857142858
      - 2.2045977011494253
      - 0.7422600619195047
      - 0.542420814479638
    ignore_index: -1
    freeze_backbone: false
    freeze_decoder: false
    model_factory: EncoderDecoderFactory
    tiled_inference_parameters:
      h_crop: 224
      h_stride: 192
      w_crop: 224
      w_stride: 192
      average_patches: true

**Comment by Joao-L-S-Almeida** [comment]
Hi, @danjac94. Could your provide more information about the test case you want to develop ? It would be helpful if you provide your complete yaml. 
I particularly don't see a great advantage in training it from scratch (using `backbone_pretrained` as `False`) in a supervised way. It could be very costing and require a large dataset to produce something reasonable. 
If you want to update the backbone, I would recommend you, in addition to using `backbone_freeze: false`, to look for [LoRA](https://github.com/IBM/terratorch/blob/86224762d4b2128f2673a5eb5bc95d2b2f911063/tests/test_encoder_decoder_model_factory.py#L401), which could facilitate it.

**Comment by danjac94** [comment]
Hi @Joao-L-S-Almeida , thank you very much for your reply. 

The first configuration that I posted was a mistake. I was trying something with Swin-B and didn't realize that i copy the wrong config. 

Exactly, I tried setting backbone_pretrained: False, backbone_freeze: False, and decoder_freeze: False as well. My main concern was that the first results seemed a bit weird when compared to when I used the pre-trained weights.

         	        mAcc	mJaccard	    mFscore
Prithvi 100M	0.9275	0.8535	    0.9275
(1) ViT base     0.9349	0.8683	    0.9349

Only when I use backbone_freeze: True, decoder_freeze: False and backbone_pretrained: False  the results changed to
(2) ViT base     0.9036	0.8132	    0.9036

I guess that the second result can be acceptable, if is the trained from scratch without using the pretrained weights, right? 

For the result (2) I use the following configuration 

# lightning.pytorch==2.1.1
seed_everything: 0
trainer:
  accelerator: auto
  strategy: auto
  devices: auto
  num_nodes: 1
  precision: 16-mixed
  logger: True
  callbacks:
    - class_path: RichProgressBar
    - class_path: LearningRateMonitor
      init_args:
        logging_interval: epoch

  max_epochs: 100
  log_every_n_steps: 50
  enable_checkpointing: true
  default_root_dir: output/scratchvit

data:
  class_path: GenericNonGeoSegmentationDataModule
  init_args:
    batch_size: 4
    num_workers: 2
    dataset_bands:
      - BLUE
      - GREEN
      - RED
      - NIR_NARROW
      - SWIR_1
      - SWIR_2
    output_bands:
      - BLUE
      - GREEN
      - RED
      - NIR_NARROW
      - SWIR_1
      - SWIR_2
    rgb_indices:
      - 2
      - 1
      - 0
    train_data_root: /dataset_stride/tiles/train
    train_label_data_root: dataset_stride/masks/train
    val_data_root: /dataset_stride/tiles/val
    val_label_data_root: /dataset_stride/masks/val
    test_data_root: /dataset_stride/tiles/val
    test_label_data_root: /dataset_stride/masks/val
    img_grep: "*_tile.tif"
    label_grep: "*_mask.tif"
    means:
      - 1290.95483036
      - 1478.05996813
      - 1382.05911294
      - 2636.1801427
      - 1981.83784814
      - 1511.65113143
    stds:
      - 18.75570915
      - 365.19936594
      - 383.25120861
      - 915.52461101
      - 646.74120943
      - 468.31984067
    num_classes: 5
    train_transform:
      - class_path: albumentations.D4
      - class_path: ToTensorV2
    reduce_zero_label: True
    no_label_replace: -1
    no_data_replace: 0

model:
  class_path: terratorch.tasks.SemanticSegmentationTask
  init_args:
    model_args:
      decoder: UperNetDecoder
      backbone_pretrained: False
      backbone: prithvi_eo_v1_100
      backbone_img_size: 256
      decoder_scale_modules: True
      # decoder_in_channels: 1024
      decoder_channels: 256
      # backbone_in_channels: 6
      backbone_model_bands:
        - BLUE
        - GREEN
        - RED
        - NIR_NARROW
        - SWIR_1
        - SWIR_2
      # backbone_out_indices:
          # - 5
          # - 11
          # - 17
          # - 23
      # num_frames: 1
      num_classes: 5
      head_dropout: 0.1
      head_channel_list:
        - 256
      necks:
        - name: SelectIndices
          indices: [2, 5, 8, 11] # 100M models
#          indices: [5, 11, 17, 23] # 300M models
#          indices: [7, 15, 23, 31] # 600M models
        - name: ReshapeTokensToImage
    loss: ce
    class_weights:
      - 1.2232142857142858
      - 1.2232142857142858
      - 2.2045977011494253
      - 0.7422600619195047
      - 0.542420814479638
    ignore_index: -1
    freeze_backbone: true
    freeze_decoder: false
    model_factory: EncoderDecoderFactory
    tiled_inference_parameters:
      h_crop: 224
      h_stride: 192
      w_crop: 224
      w_stride: 192
      average_patches: true
    
optimizer:
  class_path: torch.optim.AdamW
  init_args:
    lr: 1.5e-5
    weight_decay: 0.05
lr_scheduler:
  class_path: ReduceLROnPlateau
  init_args:
    monitor: val/loss

**Comment by Joao-L-S-Almeida** [comment]
Hi, @danjac94 . The comparison between (1) and (2) makes sense for me, since the results become worse when you don't update the backbone, that means, when your only option to train the model is to update a smaller number of parameters in the decoder and head. In anyway, it's interesting to see that the effect of the updating the backbone isn't that great. 
However, seeing that the pretrained backbone (first line, right ?) has a negative effect over the final results it's something unexpected. Maybe it's the fact you're training it for a larger than usual number of epochs could explain your model being more specialized and more effective, but it's interesting in anyway.

**Comment by Joao-L-S-Almeida** [comment]
It would also be interesting to see some visual comparison, if you could share it.

**Comment by Joao-L-S-Almeida** [comment]
It sound like it was solved. So closing the the issue.
